data_safe_haven/types/enums.py

```diff
@@ -114,6 +114,12 @@ class PermittedDomains(tuple[str, ...], Enum):
     SOFTWARE_REPOSITORIES_R = ("cran.r-project.org",)
     SOFTWARE_REPOSITORIES = SOFTWARE_REPOSITORIES_PYTHON + SOFTWARE_REPOSITORIES_R
     UBUNTU_KEYSERVER = ("keyserver.ubuntu.com",)
+    UBUNTU_SNAPCRAFT = (
+        "api.snapcraft.io",
+        "*.snapcraftcontent.com",
+        "dashboard.snapcraft.io",
+        "login.ubuntu.com",
```

Comment:
Are we sure that none of these endpoints allow upload?

Reply:
Not currently sure.

Snapcraft requires another endpoint, hopefully that is the one used for publishing 🤞.

Reply:
Doesn't seem possible to login, at least
```none
dshadmin@shm-lincolnshire-sre-morcilla-vm-workspace-01:~$ snap login
Personal information is handled as per our privacy notice at
https://www.ubuntu.com/legal/dataprivacy/snap-store

Email address: mcraddock@turing.ac.uk
Password of "mcraddock@turing.ac.uk": 
error: access denied (try with sudo)
dshadmin@shm-lincolnshire-sre-morcilla-vm-workspace-01:~$ sudo snap login
Personal information is handled as per our privacy notice at
https://www.ubuntu.com/legal/dataprivacy/snap-store

Email address: mcraddock@turing.ac.uk
Password of "mcraddock@turing.ac.uk": 
error: cannot persist login information: mkdir /home/dshadmin/.snap: permission denied
```

Reply:
> `error: cannot persist login information: mkdir /home/dshadmin/.snap: permission denied`

That sounds like a local problem though, rather than with connecting to snapcraft/the snap store.

Reply:
I agree, but haven't found a way to get it to work yet

Reply:
```none
dshadmin@shm-lincolnshire-sre-morcilla-vm-workspace-01:~$ snapcraft login
2024/05/30 14:08:50.613417 cmd_run.go:1129: WARNING: cannot create user data directory: cannot create "/home/dshadmin/snap/snapcraft/11772": mkdir /home/dshadmin/snap/snapcraft: permission denied
cannot create user data directory: /home/dshadmin/snap/snapcraft/11772: Permission denied
dshadmin@shm-lincolnshire-sre-morcilla-vm-workspace-01:~$ sudo snapcraft login
snapcraft internal error: NoKeyringError('No keyring found to store or retrieve credentials from.')                                                            
Full execution log: '/root/.local/state/snapcraft/log/snapcraft-20240530-140859.093023.log' 
```

Reply:
Nope - and I can create things with `sudo`

```none
dshadmin@shm-lincolnshire-sre-morcilla-vm-workspace-01:~$ mount | grep home
shmlincosremorciuserdata.file.core.windows.net:/shmlincosremorciuserdata/home on /home type nfs4 (rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,nconnect=4,timeo=600,retrans=2,sec=sys,clientaddr=10.2.2.4,local_lock=none,addr=10.2.1.20,_netdev)
```

Reply:
Do the leading directories exist? `mkdir` won't create directories without the `-p` flag.

Reply:
Attempting snap login as root - looks a little more promising.

Wondering if ds

```none
root@shm-lincolnshire-sre-morcilla-vm-workspace-01:~# snap login
Personal information is handled as per our privacy notice at
https://www.ubuntu.com/legal/dataprivacy/snap-store

Email address: mcraddock@turing.ac.uk
Password of "mcraddock@turing.ac.uk": 
error: cannot get snap access permission from store: Post
       "https://dashboard.snapcraft.io/dev/api/acl/": net/http: request
       canceled while waiting for connection (Client.Timeout exceeded while
       awaiting headers) (see 'snap help login')
```

Reply:
So because we block `dashboard.snapcraft.io` login is impossible?

Reply:
We don't currently block dashboard snapcraft. The issue there was the that the DNS server had changed IP address, so nothing at all was accessible.

We can now login. `snap` itself doesn't do any unloading. For that you need to install `snapcraft`. So far, I can't get `snapcraft` to build a snap to see if it can upload it 

```none
root@shm-lincolnshire-sre-chorizo-vm-workspace-01:~/snap/hithere/snap# snapcraft
craft-providers error: Failed to launch instance 'base-instance-snapcraft-buildd-base-v71--d799df4d4588b4abe861'.
* Command that failed: "lxc --project snapcraft launch craft-com.ubuntu.cloud-buildd:core22 local:base-instance-snapcraft-buildd-base-v71--d799df4d4588b4abe861 --config 'raw.idmap=both 0 0' --config security.syscalls.intercept.mknod=true --config user.craft_providers.status=STARTING --config user.craft_providers.timer=2024-06-06T09:34:24.055845+00:00 --config user.craft_providers.pid=19145"
* Command exit code: 1
* Command output: b'Creating base-instance-snapcraft-buildd-base-v71--d799df4d4588b4abe861\n'
* Command standard error output: b'Error: Failed instance creation: Failed getting remote image info: Failed getting image: Failed parsing stream: Get "https://cloud-images.ubuntu.com/buildd/releases/streams/v1/index.json": lookup cloud-images.ubuntu.com on 127.0.0.53:53: no such host\n'                                              
Full execution log: '/root/.local/state/snapcraft/log/snapcraft-20240606-093423.293280.log' 
```

`cloud-images.ubuntu.com` is blocked, so looks like that stops it.

Looks like `snapcraft` also requires http://upload.apps.ubuntu.com:443/, according to https://snapcraft.io/docs/network-requirements

Reply:
Latest: can't find a way to use the snapcraft route to upload anything. Login is achievable (at least when you are root), but building a snap won't work. The process starts by building a VM using either `multipass` or `lxd`, and neither works as the required base images cannot be downloaded. 

Unclear if it's possible to upload any other way.

Reply:
@craddm Did you try this _with_ snapcraft and other deps installed?

It might be tricky to test. Not sure if we could do something like build the snap outside the TRE then move to the TRE just to test uploading.
What I'm worried about is, could you mimic snapcraft (presumably there is some API endpoint to upload data) or ingress the tools necessary to upload.

I think we want to check that you "cannot" upload and not that it is difficult to get the normal prerequisites.